_posts/2023-09-15-kafka-producer-client-essentials.md

```diff
@@ -0,0 +1,322 @@
+
+Here's a breakdown of retriable and non-retriable errors that the client handles:
+
+**Non-retriable errors caught by the producer client**
+
+* `InterruptedException`: This error occurs when the current thread is interrupted while paused. 
+Interruption typically happens during producer shutdown or when stopping its operation. 
+The exception is rethrown as a `RuntimeException`, which ultimately terminates the producer.
+
+* `IllegalArgumentException`: This error is thrown when the producer receives invalid or inappropriate arguments. 
+For instance, it can be triggered if essential details like the topic are missing.
+
+* `UnsupportedOperationException`: This error is raised when an operation is not supported or when a method is not implemented as expected. 
+For instance, it can be triggered if you attempt to use an unsupported producer configuration or invoke a method that the `KafkaProducer` class does not support.
+
+**Retriable errors caught by the producer client**
+
+`RetriableException`: This type of error is thrown for any exception that implements the `RetriableException` interface, as provided by the Kafka client library.
+
+## Tuning your producer
+
+The example Kafka producer client in this post serves as a foundation. 
+Feel free to build on it.
+For instance, you might want to add custom functionality for integration with your preferred logging framework.
+
+You might also want to explore how to expand and improve on other aspects of your client through configuration:
+
+**Implementing security**
+
+Implement security to establish a secure connection using authentication and authorization mechanisms when connecting to the Kafka cluster. 
+For example, you can set up TLS authentication for external clients in your Strimzi environment and add the TLS certificates to your client configuration. 
+
+You can use configuration providers to load configuration, including secrets, from external sources.
+For more information see the [Strimzi documentation](https://strimzi.io/docs/operators/latest/deploying#assembly-loading-config-with-providers-str). 
+
+> Configure the security protocol used by your client application to match the protocol configured on a Kafka broker listener. 
+
+**Improving data durability** 
+  
+Specify `acks=all` (default) in your producer configuration so that all in-sync topic replicas acknowledge successful message delivery. 
+Or configure `transaction` properties in your brokers and producer client application to ensure that messages are processed in a single transaction.
+
+**Boosting performance** 
+  
+Optimize your producer for high message throughput and low latency. 
+Use the `linger.ms` and `batch.size` configuration properties to batch more messages into a single produce request for higher throughput. 
+Improve throughput of your message requests by using the `delivery.timeout.ms` property to adjust the maximum time to wait before a message is delivered and completes a send request. 
```

Comment:
Didn't we had a blog post about performance tuning? Maybe you can link to it?

Reply:
I linked to the tuning producers doc.